### ISSUE TITLE
reproducing the paper's best results

### ISSUE THREAD

**Issue by Kyubyong** [UKPLab/sentence-transformers#50]
I've tried to replicate the paper. For `bert-base-nli-mean-tokens`, the model which was trained from scratch with your code reached 74.71 of cosine-similarity on the sts-test set. It is way too low compared to the score on the paper. Any thoughts?

**Comment by chaochaoSZ** [comment]
same results with me. It is not as good as the paper saying. What tricks we don't find and not show in the paper and code.

**Comment by nreimers** [comment]
Hi,
the problem occurred when I updated huggingface pytorch-transformers to version 1.x (which was with version 2.0 of sentence-transformers): Performances magically dropped, even though the setup was as before. 

I performed extensive debugging, copying old code from huggingface, but sadly never found a way to fix it. Interesting, when loading the weights trained with the old huggingface code, the same performances were still achieved. So something must have changed in the training procedure of huggingface code that leads to this inferior performance with version 1 of pytorch-transformers. Maybe the optimizer code is a bit different?

I was not the only person affected by this, but several people mentioned this in the huggingface repo, that they now achieve slightly worse performances. The reason is unclear.

I will soon be able to update pytorch-transformers to version 2. Maybe the issue is resolved in that version? Who knows.

If you like to reproduce the old sts experiment scores, I recommend to use the older versions of this repository, one that uses pytorch-transformers 0.x version.

Best regards
Nils Reimers

**Comment by Kyubyong** [comment]
@nreimers Thanks so much. At first, I thought it may be different a bit wildly depending on the seed. But finally, it seems that it has much to do with the validation metric. (https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/EmbeddingSimilarityEvaluator.py#L25).

The following is the results from different main similarity functions.

# SimilarityFunction.COSINE: 76.20 (Cosine.Spearman) | 76.38 (Euclidean.Spearman)
# SimilarityFunction.EUCLIDEAN: 76.45 (Cosine.Spearman) | 76.60 (Euclidean.Spearman)
# SimilarityFunction.MANHATTAN: 76.70 (Cosine.Spearman) | 76.83 (Euclidean.Spearman)
# SimilarityFunction.DOT_PRODUCT: 77.10 (Cosine.Spearman) | 77.18 (Euclidean.Spearman)
# None (Default. Maximum value): 74.71 (Cosine.Spearman) | 75.03 (Manhattan.Spearman)

Strangely, the default one is the worst, and the others are better than the default one. Still not sure why this happens or this is always the case irrespective of the seeds.

**Comment by nreimers** [comment]
Hi @Kyubyong 
the default one returns the maximum of cosine/euclidean/manhatten/dot-product:
```
elif self.main_similarity is None:
            return max(eval_spearman_cosine, eval_spearman_manhattan, eval_spearman_euclidean, eval_spearman_dot)
```

I think in your case it has to do with the random seed. Performances differences due to the random seed can be quite large, especially if you have a complex model like BERT and a small dataset like the STS dataset.

Best
Nils Reimers

**Comment by K-Mike** [comment]
Hi @nreimers,

What params did you use for SentEval?

**Comment by nreimers** [comment]
Hi @K-Mike 
I used the params from the SentEval paper (which are different than the default parameters in the git repository).

I used this parameters:
```
params = {'task_path': PATH_TO_DATA, 'usepytorch': True, 'kfold': 10}
params['classifier'] = {'nhid': 0, 'optim': 'adam', 'batch_size': 64,
                                 'tenacity': 5, 'epoch_size': 4}
```
See:
https://github.com/facebookresearch/SentEval
=> "To produce results that are comparable to the literature, use the default config:"

**Comment by K-Mike** [comment]
@nreimers I continue to reproduce the paper's result.
I got 'bert-base-cased' and tried simple average, my result

MR | CR | SUBJ | MPQA | SST2 | SST5 | TREC | MRPC
81.81 | 87.52 | 95.26 | 88.55 | 86.93 | 47.23 | 87.78 | 74.51

in the article:
MR | CR | SUBJ  | MPQA |  SST |  TREC  | MRPC
78.66 |  86.25  | 94.37 |  88.66  | 84.40  | 92.8  | 69.45

Could you show the code?

**Comment by nreimers** [comment]
Hi @K-Mike 
I used in the paper bert-as-a-service with mean-pooling. Here is the code I used:

```
from __future__ import absolute_import, division, unicode_literals

import sys
import io
import numpy as np
import logging

import os
from bert_serving.client import BertClient

# Set PATHs
PATH_TO_SENTEVAL = '../'
PATH_TO_DATA = '../data'

# import SentEval
sys.path.insert(0, PATH_TO_SENTEVAL)
import senteval

# SentEval prepare and batcher
def prepare(params, samples):
    pass

bc = BertClient(ip='localhost', check_length=False)
def batcher(params, batch):
    sentences = []
    for sample in batch:
        untoken = ' '.join(sample).lower()
        if untoken == '':
            untoken = '-'

        sentences.append(untoken)
    return bc.encode(sentences)


# Set params for SentEval
#params_senteval = {'task_path': PATH_TO_DATA, 'usepytorch': True, 'kfold': 5}
#params_senteval['classifier'] = {'nhid': 0, 'optim': 'rmsprop', 'batch_size': 128, 'tenacity': 3, 'epoch_size': 2}

# Parameters suggested by Readme & https://github.com/facebookresearch/SentEval/issues/43
params_senteval = {'task_path': PATH_TO_DATA, 'usepytorch': True, 'kfold': 10}
params_senteval['classifier'] = {'nhid': 0, 'optim': 'adam', 'batch_size': 64, 'tenacity': 5, 'epoch_size': 4}

# Set up logger
logging.basicConfig(format='%(asctime)s : %(message)s', level=logging.DEBUG)

if __name__ == "__main__":
    se = senteval.engine.SE(params_senteval, batcher, prepare)
    transfer_tasks = ['MR', 'CR', 'SUBJ', 'MPQA', 'SICKEntailment', 'SST2', 'TREC', 'MRPC']
    results = se.eval(transfer_tasks)
    print(results)
```


As I learned later (pointed out in one of the issues here): Bert-as-a-service interprets mean-pooling a bit differently.
```
In the default strategy REDUCE_MEAN, I take the second-to-last hidden layer of all of the tokens in the sentence and do average pooling.
```

This might be the cause of the differences? Maybe taking only the last layer and perform mean pooling might be better than the REDUCE_MEAN pooling from bert-as-a-service? Would be interesting to see which is better.

Best 
Nils Reimers
